text/0018-bee-bundle/0018-bee-bundle.md

```diff
@@ -0,0 +1,281 @@
++ Feature name: `bee-bundle`
++ Start date: 2019-10-02
++ RFC PR: [iotaledger/bee-rfcs#18](https://github.com/iotaledger/bee-rfcs/pull/18)
++ Bee issue: [iotaledger/bee#63](https://github.com/iotaledger/bee/issues/63)
+
+# Summary
+
+This RFC is based on [`transaction-module`]().
+<!-- TODO FIx link to RFC#3 -->
+
+The smallest communication unit in the IOTA protocol is the transaction. Everything, including payment settlements and/or plain data, is propagated through the IOTA network in transactions.
+
+A transaction is `2673` trytes and the part available to the user is `2187` trytes. This part holds a signature in case of a payment settlement and plain data otherwise. Since it has a limited size, a user often needs more than one transaction to fulfil their operation, for example signatures with security level `2` or `3` don't fit in a single transaction and user-provided data may exceed the allowance so they need to be fragmented across multiple transactions. Moreover, a value transaction doesn't make sense on its own because it would change the total amount of the ledger so it has to be paired with other complementary transactions that together will balance the total value to zero.
+
+For these reasons, transactions have to be processed as a whole, in groups called bundles. A bundle is an atomic operation in the sense that either all or none of its transactions are accepted by the network. Even single transactions are propagated through the network within a bundle making it the only confirmable communication unit of the IOTA protocol.
+
+This RFC proposes ways to create and manipulate a bundle and describe the associated algorithms.
+
+Useful links:
+- [What is a bundle?
+](https://docs.iota.org/docs/getting-started/0.1/introduction/what-is-a-bundle)
+- [Bundles and transactions
+](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+
+# Motivation
+
+<!-- TODO -->
+
+# Detailed design
+
+<!-- TODO -->
+
+## Bundle and BundleBuilder
+
+Transactions are final and bundles, essentially being arrays of transactions, are also final. Once a bundle is created and validated, it shouldn't be tempered. For this reason we have a `Bundle` type and a `BundleBuilder` type. An instantiated `Bundle` object represents a syntactically and semantically valid IOTA bundle and a `BundleBuilder` is the only gateway to a `Bundle` object.
+
+### Bundle
+
+As bundles are final, they shouldn't be modifiable outside of the scope of the bundle module.
+
+A bundle can simply be represented as an array of transactions:
+
+```rust
+struct Bundle {
+    transactions: Vec<Transaction>
```

Comment:
There is a natural order. Each transaction has a `current_index` and a `last_index`. So transactions in a bundle should be ordered from `0` to `last_index`.

> Is there a chance that transactions arrive out of order

Yes definitely, but we don't really care because:
- client-side: we add transaction one at a time
- server-side: we process bundle only when we know all the transactions are present

Reply:
Yes, we might receive the transaction out of order. I think stay in vector makes sense. Otherwise, we shouldn't just use `HashMap` and should build our own linked list furthermore.